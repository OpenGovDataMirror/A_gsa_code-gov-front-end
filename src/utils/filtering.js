export function isChecked(item) {
  return typeof item === 'object' && item.checked
}