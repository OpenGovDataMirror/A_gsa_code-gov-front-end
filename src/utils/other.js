/* global PUBLIC_PATH */
/* global SITE_CONFIG */
import { find, lower, startsWith, trim } from '@code.gov/cautious'
import get from 'lodash.get'

export const falses = [undefined, null, 'null', 'None', 'Null', 'NULL', '', 'False', 'false']

export function isFalse(input) {
  return falses.includes(input)
}

export function pathMatch(url, match) {
  return (
    startsWith(url, `./${match}`) || startsWith(url, `/${match}/`) || startsWith(url, `${match}/`)
  )
}

export function updatedRecursivePath(thing) {
  if (typeof thing === 'object') {
    for (const key in thing) {
      const subvalue = thing[key]
      if (typeof subvalue === 'string') {
        // eslint-disable-next-line no-use-before-define
        thing[key] = adjustAssetPath(subvalue)
      }
    }
  }
  return thing
}

export function adjustAssetPath(thing) {
  if (pathMatch(thing, 'assets')) {
    const pattern = /.?\/?assets\//
    const newAssetPath = `${PUBLIC_PATH}assets/`
    return thing.replace(pattern, newAssetPath)
  }

  return updatedRecursivePath(thing)
}

export function getConfigValue(path) {
  if (SITE_CONFIG) {
    let value = get(SITE_CONFIG, path)
    if (typeof value === 'string') {
      try {
        value = adjustAssetPath(value)
      } catch (error) {
        console.error("couldn't check if the following started with assets", [value])
      }
    } else if (Array.isArray(value)) {
      value = value.map(item => {
        if (typeof item === 'object') {
          for (const subkey in item) {
            const subvalue = item[subkey]
            if (typeof subvalue === 'string') {
              item[subkey] = adjustAssetPath(subvalue)
            }
          }
        }
        return item
      })
    }
    if (!value) {
      console.warn(`We weren't able to find the value for ${path} in your code-gov-config.json file.
      You can consult examples under the folder found at config/site/examples
      or consult our developer documentation here: https://open.gsa.gov/api/codedotgov/`)
    }
    return value
  }
  return null
}

export function normalize(input) {
  return trim(lower(input))
}

export function setScrollDepth(scrollDepth) {
  if (document && document.documentElement && document.documentElement.scrollTop) {
    document.documentElement.scrollTop = scrollDepth
  }
  if (document.body && document.body.scrollTop) {
    document.body.scrollTop = scrollDepth
  }
}

/* runs when each page component is loaded */
export function refreshView() {
  window.scrollTo(0, 0)
  setScrollDepth(0)
  if (document.activeElement) {
    document.activeElement.blur()
  }
}

export function scrollToTopOfResults() {
  console.log('starting scrollToTopOfResults')
  const headerHeight = document.querySelector('.menu-banner-header').clientHeight
  const bannerHeight = document.querySelector('.usa-section').clientHeight
  const scrollDepth = headerHeight + bannerHeight
  const scrollOptionsSupported = document.documentElement.style.scrollBehavior !== undefined
  if (scrollOptionsSupported) {
    window.scrollTo({
      behavior: 'smooth',
      left: 0,
      top: scrollDepth
    })
  } else {
    window.scrollTo(0, scrollDepth)
  }
  setScrollDepth(scrollDepth)
}

/* gets a set of values given a path */
export function getSet(items, path) {
  const results = new Set()
  if (Array.isArray(items) && typeof path === 'string') {
    items.forEach(item => {
      const value = get(item, path)
      if (value && value !== '') {
        if (Array.isArray(value)) {
          value.forEach(subitem => {
            if (isFalse(subitem) === false) {
              results.add(subitem)
            }
          })
        } else if (isFalse(value) === false) {
          results.add(value)
        }
      }
    })
  }
  return results
}

export function getLowerSet(items, path) {
  const results = new Set()
  if (Array.isArray(items) && typeof path === 'string') {
    items.forEach(item => {
      const value = get(item, path)
      if (value && value !== '') {
        if (Array.isArray(value)) {
          value.forEach(subitem => {
            if (isFalse(subitem) === false) {
              results.add(subitem.toLowerCase())
            }
          })
        } else if (isFalse(value) === false) {
          results.add(value.toLowerCase())
        }
      }
    })
  }
  return results
}

export function getFilterData(key, path, currentSearchResults, filters) {
  if (currentSearchResults && filters && filters[key]) {
    const names = getLowerSet(currentSearchResults.repos, path)
    return filters[key].filter(
      ({ name, value }) => names.has(normalize(name)) || names.has(normalize(value))
    )
  }
}

export function hasLicense(repo) {
  return (
    repo.permissions &&
    Array.isArray(repo.permissions.licenses) &&
    repo.permissions.licenses.length > 0
  )
}

export function now() {
  return new Date().getTime()
}

export function getFilterValuesFromParamsByCategory(params = {}, category) {
  console.log('other-getFilter: category', category)
  return params.filters
    .filter(entry => entry.category === category)
    .map(entry => entry.value)
    .map(value => (typeof value === 'string' ? value.toLowerCase().trim() : value))
}

export function getFilterTags(params, filters) {
  return params.filters
    .map(({ category, modified, value }) => {
      const normalizedValue = value.toLowerCase()
      const found = find(
        get(filters, category),
        item => item.value.toLowerCase() === normalizedValue
      )
      let title = 'loading'
      if (found) {
        /* eslint-disable */
        if (found.name) title = found.name
        if (found.value) value = found.value
        /* eslint-enable */
      }
      return { category, modified, value, title }
    })
    .sort((a, b) => Math.sign(a.modified - b.modified))
}

export function loadScript(src) {
  console.log('starting loadScript with', src)
  return new Promise((resolve, _reject) => {
    const script = document.createElement('script')
    script.src = src
    script.onload = resolve
    document.body.appendChild(script)
    console.log('appended:', script, 'to body')
  })
}

export function fillFilters(keys, params, result) {
  keys.forEach(key => {
    if (Array.isArray(params[key])) {
      params[key].forEach(value => {
        result.filters.push({
          category: key,
          value,
          modified: now()
        })
      })
    }
  })
}

export function onHomePage() {
  return window.location.pathname === PUBLIC_PATH
}

/* I'd prefer to use fetch but IE polyfilling is complicated */
export function getText(url) {
  return new Promise(resolve => {
    const xhr = new XMLHttpRequest()
    xhr.onreadystatechange = function() {
      if (xhr.readyState === 4) {
        resolve(xhr.response)
      }
    }
    xhr.open('GET', url, true)
    xhr.send('')
  })
}

export function getJSON(url) {
  return getText(url).then(JSON.parse)
}

// https://stackoverflow.com/questions/1500260/detect-urls-in-text-with-javascript
export function urlify(text) {
  const urlRegex = /(https?:\/\/[^\s']+)/g
  return text.replace(urlRegex, url => `<a href="${url}" target="_blank">${url}</a>`)
}

export function prettify(text) {
  // convert (1) cost: to \n\t(1) <b>cost</b>:
  const bulletRegex = /\(\d{1,2}\) [A-Za-z]{1,25}:/g
  return urlify(text).replace(
    bulletRegex,
    match => `<br/>${match.replace(/[A-Za-z]{1,25}/, name => `<b>${name}</b>`)}`
  )
}

export const isHomepage = window.location.pathname === PUBLIC_PATH
