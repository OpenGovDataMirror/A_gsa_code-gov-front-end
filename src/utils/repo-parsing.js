import get from 'lodash.get'
import { filter } from '@code.gov/cautious'

export function getDisplayTextForUsageType(repo) {
  let text = ''
  const usageType = get(repo, 'permissions.usageType')
  if (usageType === 'openSource') {
    text = 'Open Source'
  } else if (usageType === 'governmentWideReuse') {
    text = 'Gov-wide Reuse'
  } else if (usageType) {
    text = usageType
  } else {
    text = 'unknown'
  }
  return text
}

export function getLicenseName(repo) {
  return get(repo, 'permissions.licenses[0].name')
}

export function getLaborHours(repo) {
  try {
    const laborHours = Number(get(repo, 'contact.laborHours'))
    if (laborHours > 0) {
      return laborHours
    }
  } catch (error) {
    console.warn(error)
  }
}

export function parseLanguages(repo) {
  const languages = get(repo, 'languages')
  if (Array.isArray(languages)) {
    return filter(languages, Boolean)
  }
  return []
}

export function parseEmail(repo) {
  const email = get(repo, 'contact.email')
  if (email) {
    return email
  }
}

export function parseRepositoryURL(repo) {
  let url = get(repo, 'repositoryURL')

  // some repos return 'null'
  if (!url || url === 'null') {
    return ''
  }

  if (url.startsWith('git://github.com/')) {
    url = url.replace('git://github.com/', 'https://github.com/')
  } else if (url.startsWith('git@github.com:')) {
    url = url.replace('git@github.com:', 'https://github.com/').replace('.git', '')
  } else if (url.startsWith('https://github.com') && url.endsWith('.git')) {
    url = url.replace('.git', '')
  }

  return url
}

export function parseTags(repo) {
  try {
    return get(repo, 'tags').filter(Boolean)
  } catch (error) {
    return []
  }
}

export function getDate(item) {
  return get(item, 'ghUpdatedAt') || get(item, 'date.lastModified')
}

export function getLastModifiedDateString(repo) {
  try {
    const dateLastModified = getDate(repo)
    if (dateLastModified) {
      return new Date(dateLastModified).toLocaleDateString('en-US')
    }
  } catch (error) {
    console.warn(error)
  }
}
