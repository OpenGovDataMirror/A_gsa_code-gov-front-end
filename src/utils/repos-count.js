export function getReposCount(total, messages, cssClasses = '') {
  let textContent = messages.default

  const len = Object.keys(messages).length - 1
  if (total >= len) {
    textContent = Object.values(messages)[len - 1]
  } else {
    Object.values(messages).forEach((item, key) => {
      if (key === total) {
        textContent = item
      }
    })
  }

  textContent = textContent.replace('[VALUE]', total)
  return `<p class="${cssClasses}">${textContent}</p>`
}
