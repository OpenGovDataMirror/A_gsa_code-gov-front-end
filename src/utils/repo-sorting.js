import { getDate } from './repo-parsing'

export function clean(string) {
  try {
    return string
      .toLowerCase()
      .replace(/-/g, ' ')
      .trim()
  } catch (error) {
    return string
  }
}

export function getTime(date) {
  try {
    if (date) {
      return new Date(date).getTime()
    }
    return -10e10
  } catch (error) {
    return -10e10
  }
}

export function sortByBestMatch(a, b) {
  return Number(a.searchScore) > Number(b.searchScore) ? -1 : 1
}

export function sortByDataQuality(a, b) {
  return Number(a.score) > Number(b.score) ? -1 : 1
}

export function sortByDate(a, b) {
  try {
    const aDate = getDate(a)
    const bDate = getDate(b)
    const aTime = getTime(aDate)
    const bTime = getTime(bDate)
    return Math.sign(bTime - aTime) || 0
  } catch (error) {
    console.error(error)
    return 0
  }
}

export function sortByName(a, b) {
  return clean(a.name) > clean(b.name) ? 1 : -1
}
