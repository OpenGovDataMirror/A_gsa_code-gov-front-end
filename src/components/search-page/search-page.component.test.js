import React from 'react'
import { shallow } from 'enzyme'

import { testRenderText, testRenderList, testRenderEmpty } from 'testUtils/render'
import { refreshView, scrollToTopOfResults } from 'utils/other'
import RepoCard from 'components/repo-card'
import SearchPage from 'components/search-page/search-page.component'

jest.mock('utils/other')

const props = {
  total: 2,
  searchParams: { query: 'test-params-query' },
  query: 'test-query',
  filteredResults: [{ repoId: 'repo-1' }, { repoId: 'repo-2' }],
  boxes: ['box-1', 'box-2'],
  sortOptions: ['sort-1', 'sort-2'],
  onSortChange: jest.fn(),
  filterTags: ['filter-1', 'filter-2'],
  onFilterTagClick: jest.fn(),
  filterData: false,
  selectedPage: 1,
  selectedPageSize: 2,
  saveFilterData: jest.fn(),
  onFilterBoxChange: jest.fn(),
  updatePage: jest.fn()
}

let wrapper
let instance
describe('components - SearchPage', () => {
  beforeEach(() => {
    wrapper = shallow(<SearchPage {...props} />)
    instance = wrapper.instance()
  })

  describe('componentDidMount', () => {
    it('should refresh the view', () => {
      expect(refreshView).toBeCalled()
    })

    it('should save the filter data', () => {
      expect(props.saveFilterData).toBeCalled()
    })

    it('should not save filter data if it already exists', () => {
      jest.resetAllMocks()
      wrapper.setProps({ filterData: true })
      instance.componentDidMount()
      expect(props.saveFilterData).not.toBeCalled()
    })
  })

  describe('shouldComponentUpdate', () => {
    it('should update if next props are not current props', () => {
      const nextProps = { ...props, selectedPageSize: 8 }
      expect(instance.shouldComponentUpdate(nextProps)).toBeTruthy()
    })

    xit('should not update if next props are the same as current props', () => {
      // BUG: comparing stringified next props vs non-string current props, always updates
      // (and unnecessary comparing of state when state is not used)
      expect(instance.shouldComponentUpdate(props)).toBeFalsy()
    })
  })

  describe('onFilterBoxChange', () => {
    beforeEach(() => {
      instance.onFilterBoxChange('category', 'values')
    })

    it('should scroll to the top of the results', () => {
      document.body.innerHTML =
        '<header class="usa-card__header grid-col-9"><h3 class="usa-card__heading font-heading-lg margin-top-0"><a class="project-link" href="/projects/nsf_1_ember_common_components">ember-common-components</a></h3></header>'
      expect(scrollToTopOfResults).toBeCalled()
    })

    it('should change the filter box value', () => {
      expect(props.onFilterBoxChange).toBeCalledWith('category', 'values')
    })
  })

  describe('repoCounter', () => {
    it.each`
      total        | match
      ${0}         | ${/no/i}
      ${1}         | ${/1/}
      ${9}         | ${/9/}
      ${undefined} | ${/loading/i}
    `('should render text that matches $match when the total is $total', ({ total, match }) => {
      wrapper.setProps({ total })
      testRenderText(instance.repoCounter, match)
    })

    it('should render text that matches `loading` when no `filteredResults` provided', () => {
      wrapper.setProps({ filteredResults: undefined })
      testRenderText(instance.repoCounter, /loading/i)
    })
  })

  describe('reposContainer', () => {
    it('should render a list of all `filteredResults` in `RepoCard`s', () => {
      testRenderList(instance.reposContainer, RepoCard, props.filteredResults.length)
    })

    it('should render nothing if no `filteredResults`', () => {
      wrapper.setProps({ filteredResults: undefined })
      testRenderEmpty(instance.reposContainer)
    })
  })

  describe('updatePage', () => {
    beforeEach(() => {
      instance.updatePage('page')
    })

    it('should scroll to the top of the results', () => {
      expect(scrollToTopOfResults).toBeCalled()
    })

    it('should update the page value', () => {
      document.body.innerHTML =
        '<header class="usa-card__header grid-col-9"><h3 class="usa-card__heading font-heading-lg margin-top-0"><a class="project-link" href="/projects/nsf_1_ember_common_components">ember-common-components</a></h3></header>'
      expect(props.updatePage).toBeCalledWith('page')
    })
  })

  describe('render', () => {
    it('should render correctly', () => {
      expect(wrapper).toMatchSnapshot()
    })
  })
})
