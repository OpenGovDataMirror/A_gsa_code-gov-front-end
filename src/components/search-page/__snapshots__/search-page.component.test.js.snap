// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components - SearchPage render should render correctly 1`] = `
<main
  className="search-results-content"
  id="main-content"
>
  <Connect(SiteBanner)
    title="Search Results"
  />
  <Breadcrumbs
    crumbs={
      Array [
        Object {
          "text": "Home",
          "to": "/",
        },
        Object {
          "text": "Search Results",
        },
      ]
    }
  />
  <div
    className="grid-container"
  >
    <div
      className="grid-row grid-gap"
    >
      <div
        className="margin-top-1 grid-col-12 tablet-lg:grid-col-3"
      >
        <Connect(QuickSearchBoxComponent)
          value="test-params-query"
        />
      </div>
      <p
        className="repos-count tablet-lg:grid-col-9 grid-col-12 font-heading-lg text-bold tablet:margin-y-105 margin-top-3 margin-bottom-0"
      >
        Top 2 repositories for "test-query"
      </p>
    </div>
  </div>
  <div
    className="grid-container"
  >
    <div
      className="grid-row grid-gap"
    >
      <div
        className="tablet-lg:grid-col-3 tablet-lg:margin-top-4 margin-top-3"
        id="filter-boxes-section"
      >
        <h2
          className="tablet-lg:margin-bottom-4 margin-bottom-105"
        >
          Filter
        </h2>
        <FilterBoxes
          boxes={
            Array [
              "box-1",
              "box-2",
            ]
          }
          config={
            Array [
              Array [
                "Language",
                "languages",
              ],
              Array [
                "Federal Agency",
                "agencies",
              ],
              Array [
                "Licenses",
                "licenses",
              ],
              Array [
                "Usage Types",
                "usageTypes",
              ],
            ]
          }
          onFilterBoxChange={[Function]}
        />
      </div>
      <div
        className=" tablet-lg:grid-col-9"
        id="filter-results-section"
      >
        <SortSection
          onSortChange={[MockFunction]}
          options={
            Array [
              "sort-1",
              "sort-2",
            ]
          }
        />
        <_default
          filters={
            Array [
              "filter-1",
              "filter-2",
            ]
          }
          onClick={[Function]}
        />
        <div
          className="card-list"
        >
          <div>
            <QualityPopover />
            <ul
              className="usa-card-group padding-top-3"
            >
              <RepoCardComponent
                repo={
                  Object {
                    "repoId": "repo-1",
                  }
                }
              />
              <RepoCardComponent
                repo={
                  Object {
                    "repoId": "repo-2",
                  }
                }
              />
            </ul>
          </div>
          <Pagination
            count={2}
            page={1}
            pagesize={2}
            updatePage={[Function]}
          />
        </div>
      </div>
    </div>
  </div>
</main>
`;
